system/classes/KO7/Controller/REST.php

```diff
@@ -0,0 +1,198 @@
+		];
+
+		// Copy all http header(s)
+		foreach ($response->headers()->getArrayCopy() as $header => $value)
+		{
+			$config[$header] = $value;
+		}
+
+		// Instance Response_REST and pass it to the constructor
+		parent::__construct($request, Response_REST::factory($config));
+	}
+
+	/**
+     * Automatically executed before the controller action.
+     * Evaluate Request (method, action, parameter, format)
+     *
+     * @throws HTTP_Exception
+	 */
+	public function before() : void
+	{
+		// Parent call
+		parent::before();
+
+        // Try fetching method from HTTP_X_HTTP_METHOD_OVERRIDE, otherwise use the one sent with the request
+        $this->request->method($_SERVER['HTTP_X_HTTP_METHOD_OVERRIDE'] ?? $this->request->method());
+
+		// Determine the request action from the request method, if the action/method is not allowed throw error
+        if ( ! isset(static::$_action_map[$this->request->method()]))
+        {
+            $this->response
+                ->status(405)
+                ->headers('Allow', implode(', ', array_keys(static::$_action_map)));
+        }
+        else
+        {
+            $this->request->action(static::$_action_map[$this->request->method()]);
+        }
+
+		// Initialize request parameter
+        $this->_params = $this->_parse_params();
+
+		// Get output format from route - use default if not available
+		$this->output_format = $this->request->param('format') ?: static::$default_output;
+
+		// Try initializing the formatter
+		try
+        {
+            $this->_formatter = REST_Format::factory(strtoupper($this->output_format), $this->request);
+        }
+        catch (REST_Exception $e)
+        {
+            throw HTTP_Exception::factory(500, $e->getMessage(), NULL, $e);
+        }
+	}
+
+	/**
+     * Automatically executed after the controller action.
+     *
+	 * - Adds cache and content header(s).
+     * - Formats body with given formatting method
+     * - Adds attachment header if necessary
+	 */
+	public function after() : void
+	{
+        // Parent call
+        parent::after();
+
+		// No cache / must-revalidate cache if method is not GET
+        if ( $this->request->method() !== HTTP_Request::GET)
+        {
+            $this->response->headers('cache-control', 'no-cache, no-store, max-age=0, must-revalidate');
+        }
+
+		$body = $this->response->body();
+
+        // Check if body is array, else convert it to one by creating an array with "body" as index
+		if ( ! is_array($body))
+		{
+			$body = [
+				'body' => $body
+			];
+		}
+
+		// Format body
+        $body = $this->_formatter->format($body);
+
+		// Parse and set response headers
+		$this->response->headers('content-type', $this->_content_type ?? File::mime_by_ext($this->output_format));
+		$this->response->headers('content-length', (string) strlen($body));
+
+		// Support attachment header
+		if (isset($this->_params['attachment']) && Valid::regex($this->_params['attachment'], '/^[-\pL\pN_, ]++$/uD'))
```

Comment:
it's not for controller as half this code

Reply:
What do you mean?

Reply:
logic in controller, not SOLID

Reply:
Agree, i will take a closer look into this and check if i can move those functions where they belong

Reply:
Moved stuff belonging to the response class, into there,...will move the ones belonging to the request class within this week than we should be good to go

Reply:
Moved functions and operations where it was possible to there files (Response and Request), moved some functions to already implemented fuctions and removed unneccessary properties from classes....